datasets/dataset_plugin.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+from abc import ABC
+from typing import Callable, Dict, List, Optional
+
+from datasets.context import Context
+
+from .mode import Mode
+from .program_executor import ProgramExecutor
+
+
+class DatasetPlugin(ABC):
+    """
+    All dataset plugins derive from this class.
+    To register as a dataset they must decorate themselves with or call DatasetPlugin.register_plugin()
+    """
+
+    _executor: ProgramExecutor
+    # Context -> constructor_keys -> dataset plugin
+    _plugins: Dict[Context, Dict[set[str], DatasetPlugin]] = {}
+    _META_COLUMNS = ["run_id"]
+
+    def __init__(
+        self,
+        name: str,
+        logical_key: Optional[str] = None,
+        columns=None,
```

Comment:
I think we touched on this but perhaps we should type hint on columns that it should be `Union[Iterable[str], str]`? Going to add a related comment on _get_read_columns